src/Shared/FileUtilitiesRegex.cs

```diff
@@ -34,5 +34,180 @@ internal static class FileUtilitiesRegex
+            }
+
+            bool prevCharWasSlash = true;
+            bool hasSubfolder = false;
+            for (int i = 2; i < pattern.Length; i++)
+            {
+                if (pattern[i] == Path.DirectorySeparatorChar ||
+                    pattern[i] == Path.AltDirectorySeparatorChar)
+                {
+                    if (prevCharWasSlash || hasSubfolder)
+                    {
+                        //We get here in the case of an extra slash or multiple subfolders
+                        //  Note this function is meant to mimic the UncPattern regex above.
+                        return false;
+                    }
+
+                    hasSubfolder = true;
+                    prevCharWasSlash = true;
+                }
+                else
+                {
+                    prevCharWasSlash = false;
+                }
+            }
+
+            //Valid unc patterns don't end with slashes & have at least 1 subfolder
+            return !prevCharWasSlash && hasSubfolder;
+        }
+
+        /// <summary>
+        /// Indicates whether the specified file-spec begins with "\\<server>\<path>".
+        /// </summary>
+        /// <param name="pattern"></param>
+        /// <returns></returns>
+        internal static bool DoesStartWithUncPattern(string pattern)
+        {
+            if (!MeetsUncPatternMinimumRequirements(pattern))
+            {
+                return false;
+            }
+
+            bool prevCharWasSlash = true;
+            bool hasSubfolder = false;
+
+            for (int i = 2; i < pattern.Length; i++)
+            {
+                if (pattern[i] == Path.DirectorySeparatorChar ||
+                    pattern[i] == Path.AltDirectorySeparatorChar)
+                {
+                    if (prevCharWasSlash)
+                    {
+                        //We get here in the case of an extra slash.
+                        return false;
+                    }
+
+                    hasSubfolder = true;
+                    prevCharWasSlash = true;
+                }
+                else
+                {
+                    if (hasSubfolder)
+                    {
+                        //A character after a subfolder confirms the beginning of a unc pattern
+                        return true;
+                    }
+
+                    prevCharWasSlash = false;
+                }
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Indicates whether the file-spec begins with a UNC pattern and how long the match is. -1 indicates no match.
+        /// </summary>
+        /// <param name="pattern"></param>
+        /// <returns></returns>
+        internal static int DoesStartWithUncPatternMatchLength(string pattern)
+        {
+            if (!MeetsUncPatternMinimumRequirements(pattern))
+            {
+                return -1;
+            }
+
+            bool prevCharWasSlash = true;
+            bool hasSubfolder = false;
```

Comment:
Is this a subfolder, or the share name in `\\server\share`?

Reply:
This is the share. Think this is due for a rename?

Reply:
Yeah, let's use the official names:

```abnf
UNC = "\\" host-name "\" share-name [ "\" object-name ]
host-name = "[" IPv6address "]" / IPv4address / reg-name
; IPv6address, IPv4address, and reg-name as specified in [RFC3986]
share-name = 1*80pchar
pchar = %x20-21 / %x23-29 / %x2D-2E / %x30-39 / %x40-5A / %x5E-7B / %x7D-FF
object-name = *path-name [ "\" file-name ]
path-name = 1*255pchar
file-name = 1*255fchar [ ":" stream-name [ ":" stream-type ] ]
fchar = %x20-21 / %x23-29 / %x2B-2E / %x30-39 / %x3B / %x3D / %x40-5B / %x5D-7B
/ %x7D-FF
stream-name = *schar
schar = %x01-2E / %x30-39 / %x3B-5B /%x5D-FF
stream-type = 1*schar
```

(from the [Windows Data Types](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-dtyp/cca27429-5689-4a16-b2b4-9325d93e4ba2) spec)